### ISSUE TITLE
Resolve "Ability to deep-copy a tree/forest" - [merged]

### ISSUE THREAD

**Issue by joergbrech** [grunk-dev/parametric#42]
_Merges 12-ability-to-deep-copy-a-tree-forest -> master_

Closes #12. 
Closes #13, which was needed to fix #12. #13 was fixed as a merge request !17 into this branch.

This MR provides the possibility of deep-copying a parametric tree / parametric forest while maintaining parent-child relations. 

In parametric, a user stores some nodes of a tree, e.g. inputs and outputs. Intermediate nodes need not be used by a user, they are stored in the parent-child relations of the nodes. If a user simply copies the nodes, they will internally store the parent-child relations of the copied-from nodes. Therefore, in a deep-copy, we need to recreate the parent-child relations of the input nodes for the output nodes.

This is achieved in the following way:

- The base class `DAGNode` of all nodes in a parametric tree obtain a new virtual method `clone`, which must be overridden to create a deep-copy of this node. It should recursively deep-copy all of its parents and reset the parent child relations. 
- During the recursive call, we may visit intermediate results more than one time. We have to make sure, that we do not create duplicate copies. Therefore, the `DAGNode` clone method has an optional argument, which is a map of `DAGNode` instances that have already been cloned. 
- parametric allows the use of custom `DAGNode` instances, by allowing the user to inherit `parametric::ComputeNode` which itself inherits `DAGNode`. To elevate the burden of having a user provide a custom `clone` function, a new CRTP class `template <typename Derived> class ClonableDAGNode<Derived> : public DAGNode` was introduced. Via CRTP, the `ClonableDAGNode` knows how to copy derived instances and can final-override the `clone` method of `DAGNode`. The clone method

  - checks the `cloned_node` map, if the node has already been cloned. If yes, it returns the cloned node. If no:
  - Creates a new shared_ptr from a copy of the current node.
  - Deletes the copied parent-child relations
  - For all parents of the copied-from node, it recursively calls the `clone` method and resets the parent-child relations of the copied-to node by attaching it to the cloned parent node.
  
  Previously, `impl::param_holder` and `ComputeNode` where direct child-classes of `DAGNode`. Now they inherit `ClonableDAGNode`. Since `ComputeNode` can be inherited by users, it must be a CRTP class too, so that we can properly propagate the `Derived` type to `ClonableDAGNode`. Therefore `ClonableDAGNode` has a new template parameter `Derived`.

All of this would have been enough, if the parent-child relations were only stored within a `DAGNode`. Since this is not the case, we need a redesign of `ComputeNode` to fix the redundant storage of parent-child relations. This is a bigger change and has been put into #13 and !17. 

Note to reviewer: The MR !17 was merged into this branch rather than main. This allowed me to develop more quickly, because I wouldn't have to rebase this branch on main after a merge of !17. This has the downside that this MR is rather big. But hopefully, the changes are more clear by reading the two MRs separately.

**Reviewers:** @AntonReiswich

**Approved by:** @AntonReiswich

*Migrated from GitLab: https://gitlab.dlr.de/paradigms/parametric/-/merge_requests/16*

**Comment by joergbrech** [comment]
added 1 commit

<ul><li>1c6bfe24 - implement DAGNode::clone</li></ul>

[Compare with previous version](/paradigms/parametric/-/merge_requests/16/diffs?diff_id=374667&start_sha=26df6e48dfb233966c25dd85da4b2de3bdf3592c)

**Comment by joergbrech** [comment]
added 1 commit

<ul><li>cbdf3eaa - implement DAGNode::clone</li></ul>

[Compare with previous version](/paradigms/parametric/-/merge_requests/16/diffs?diff_id=374675&start_sha=1c6bfe2481d6994937c990090fd04e63088498e8)

**Comment by joergbrech** [comment]
mentioned in issue grunk#40

**Comment by joergbrech** [comment]
added 1 commit

<ul><li>dcc63d27 - continue work on clonable trees</li></ul>

[Compare with previous version](/paradigms/parametric/-/merge_requests/16/diffs?diff_id=375562&start_sha=cbdf3eaac14a25c379c31b5d8b19006b65782b8b)

**Comment by joergbrech** [comment]
added 17 commits

<ul><li>dcc63d27...bf30ad7a - 7 earlier commits</li><li>9d10dac1 - add post_connect callback</li><li>f615a624 - documentation</li><li>222414e3 - set cxx_std_17 for unit tests</li><li>68795acb - make MSVC 2017 happy</li><li>75c7da0c - test_package: use C++17</li><li>eeda85b3 - add missing doc for param::compute_node</li><li>ea691e5c - fix void functions and add unit tests</li><li>32d87343 - make msvc 2017 happy again</li><li>bf0738e0 - allow duplicate entries in parents, childs</li><li>d9421bd1 - Merge branch &#39;13-redesign-parametric-compute-node&#39; into &#39;12-ability-to-deep-copy-a-tree-forest&#39;</li></ul>

[Compare with previous version](/paradigms/parametric/-/merge_requests/16/diffs?diff_id=377864&start_sha=dcc63d273b9aa48d72b9203a1fff5244968eec14)

**Comment by joergbrech** [comment]
marked this merge request as **ready**

**Comment by joergbrech** [comment]
requested review from @AntonReiswich

**Comment by joergbrech** [comment]
added 1 commit

<ul><li>cf920ff7 - unique childs, duplicate parents</li></ul>

[Compare with previous version](/paradigms/parametric/-/merge_requests/16/diffs?diff_id=378069&start_sha=d9421bd1c5b01053c6972c8b33870f4f9810aa3e)

**Comment by joergbrech** [comment]
**Discussion in GitLab:**

@AntonReiswich I just notices a minor problem in the interface: 

Previously, `ComputeNode` was not a template. It is now, so that the argument and result types are known and the correct type can be returned when using the members `arg` or `res`.

For the static `Action` in grunk this interface is perfectly fine. For the `DynamicAction` it could be a bit inconvenient, because the idea behind dynamic actions is that the type and number of arguments may only be known at runtime. But currently, the `Action` classes derive from `ComputeNode`, which encodes the function signature in its template arguments.

There may be a way to work around this in grunk *(I haven't had a great idea how to yet)*, but I think it might make more sense to simplify the interface of parametric here.

We could for instance provide an overload of `parametric::compute` that accepts `std::vector<param<T>>`. So this would allow connecting a vector of input parameters, all wrapping the same type to a compute node and in consequence to the outputs. In addition, we would need to provide a new overload for `arg`, which doesn't query the type of the i-th argument, but lets the user provide it:

```cpp
// use this to get the i-th argument, based on the argument type of the compute node
template <int i>
decltype(auto) arg() const { 
        assert(this->parents.size() == std::tuple_size_v<Arguments>);
        using value_type = typename std::decay_t<std::tuple_element_t<i, Arguments>>::value_type;
        return arg<i, value_type>();
}

// use this to ignore the argument type of the compute node
template <int i, typename value_type>
decltype(auto) arg() const {
        assert(i < this->parents.size());
        return dynamic_cast<impl::param_holder<value_type>&>(*(this->parents[i]));
}
```

This way, we can use it in grunk somewhat like this:

```cpp
template <>
class DynamicAction<reflect::DynamicFunction> : public parametric::ComputeNode<
                                                          DynamicAction<reflect::DynamicFunction>,
                                                          parametric::Results<std::vector<reflect::DynamicObject>>,
                                                          std::void_t // <-- doesn't matter what we put here
                                                       >
{
   // .... 

   void eval() const override 
   {
      if (auto& res = this->res<0>(); res) {

         for (size_t i = 0; i< num_inputs; ++i) {
             auto const& arg = this->arg<i, reflect::DynamicObject>();
             // ...
         }
      }
   }
}

decltype(auto) action(std::string const& id, reflect::DynamicFunction const& fun, std::vector<DynamicFeature const&> args) 
{
   // transform args vector to std::vector<param<DynamicObject>>
   // call overload of parametric::compute that accepts vector of params
}
```

I can add this now, or until you are finished with the review this far. What do you prefer?

***In GitLab by @AntonReiswich on Sep 13, 2023, 09:03 UTC:***

I think it is ok if you add it now, then you are done with it.

Actually, I think I changed my mind: Let's close this MR as is and then I will open a new MR after that. It is already large as is and I need to experiment a bit before I have a good solution.

**Comment by joergbrech** [comment]
**Discussion in GitLab:**

***In GitLab by @AntonReiswich on Sep 13, 2023, 09:02 UTC:***

Commented on [parametric/impl/core_impl.hpp line 73](https://github.com/grunk-dev/parametric/compare/26df6e48dfb233966c25dd85da4b2de3bdf3592c..cf920ff7f8cb30a2f527471fda19fc361e65549a#diff-d10ae0affcea66cef23ae61f7cc90222ba4708e04da623dde6713f24dca58291R73)

https://github.com/grunk-dev/parametric/blob/cf920ff7f8cb30a2f527471fda19fc361e65549a/parametric/impl/core_impl.hpp#L73



Why do you use this->invalidate(); now instead of invalidate();?

https://stackoverflow.com/questions/4643074/why-do-i-have-to-access-template-base-class-members-through-the-this-pointer

***In GitLab by @AntonReiswich on Sep 13, 2023, 09:16 UTC:***

Ok, cool!

**Comment by joergbrech** [comment]
**Discussion in GitLab:**

***In GitLab by @AntonReiswich on Sep 13, 2023, 09:08 UTC:***

Commented on [parametric/impl/core_impl.hpp line 116](https://github.com/grunk-dev/parametric/compare/cbdf3eaac14a25c379c31b5d8b19006b65782b8b..dcc63d273b9aa48d72b9203a1fff5244968eec14#diff-d10ae0affcea66cef23ae61f7cc90222ba4708e04da623dde6713f24dca58291L116)

https://github.com/grunk-dev/parametric/blob/dcc63d273b9aa48d72b9203a1fff5244968eec14/parametric/impl/core_impl.hpp#L116



Same questions here: Why do you use this->parents.size() instead of parents.size()?

See above, it is required when using inheritence together with class templates.

***In GitLab by @AntonReiswich on Sep 13, 2023, 09:16 UTC:***

Ok great, thanks.

**Comment by joergbrech** [comment]
**Discussion in GitLab:**

***In GitLab by @AntonReiswich on Sep 13, 2023, 09:31 UTC:***

Commented on [parametric/impl/core_impl.hpp line 51](https://github.com/grunk-dev/parametric/compare/26df6e48dfb233966c25dd85da4b2de3bdf3592c..cf920ff7f8cb30a2f527471fda19fc361e65549a#diff-d10ae0affcea66cef23ae61f7cc90222ba4708e04da623dde6713f24dca58291R51)

https://github.com/grunk-dev/parametric/blob/cf920ff7f8cb30a2f527471fda19fc361e65549a/parametric/impl/core_impl.hpp#L51



What is such a constructor used for?

This was introduced in https://gitlab.dlr.de/paradigms/parametric/-/issues/2 and https://gitlab.dlr.de/paradigms/parametric/-/merge_requests/9. It was motivated by the same concept in the std library (https://en.cppreference.com/w/cpp/utility/in_place). 

In short, `std::in_place_t` is a "tag type", whos sole purpose is to differentiate an overload via the signature. With this, we specify that the arguments shall be forwarded to a constructor and that the value shall be constructed in place without creating temporary objects and have superfluous move or copy operations. 

See also here: https://www.cppstories.com/2018/07/in-place-cpp17/

**Comment by joergbrech** [comment]
**Discussion in GitLab:**

***In GitLab by @AntonReiswich on Sep 13, 2023, 10:13 UTC:***

Thanks! I am still confused why this constructor can take variadic template parameter packs. Isn't the class impl::param_holder implemented in such a way that it can only store one std::optional value data member?

The parameters in the parameter pack are forwarded to the constructor of the held value. The constructor may accept more than one argument.

***In GitLab by @AntonReiswich on Sep 13, 2023, 10:19 UTC:***

I see, great!

**Comment by joergbrech** [comment]
**Discussion in GitLab:**

***In GitLab by @AntonReiswich on Sep 13, 2023, 10:25 UTC:***

Commented on [parametric/dag.hpp line 353](https://github.com/grunk-dev/parametric/compare/26df6e48dfb233966c25dd85da4b2de3bdf3592c..cf920ff7f8cb30a2f527471fda19fc361e65549a#diff-96aeefb0374ca00abc002562ed64c573e49b159ed8f4dc21d2f5557ff33746aeR353)

https://github.com/grunk-dev/parametric/blob/cf920ff7f8cb30a2f527471fda19fc361e65549a/parametric/dag.hpp#L340



The parents are recreated in the copied tree. What about the childs of the nodes in the copied tree? They don't seem to get recreated.

***In GitLab by @AntonReiswich on Sep 13, 2023, 10:28 UTC:***

I mean to say, that the weak pointer that connect the parent DAGNode to a child DAGNode are not set.

***In GitLab by @AntonReiswich on Sep 13, 2023, 10:29 UTC:***

Oh sorry, I see, this is done in the add_parent(...) function already.

**Comment by joergbrech** [comment]
**Discussion in GitLab:**

***In GitLab by @AntonReiswich on Sep 13, 2023, 14:49 UTC:***

Commented on [parametric/core.hpp line 162](https://github.com/grunk-dev/parametric/compare/26df6e48dfb233966c25dd85da4b2de3bdf3592c..cf920ff7f8cb30a2f527471fda19fc361e65549a#diff-092665cd611f3e8daaf379734eaf4a6a9e96a506bf09e52d584eeb76bc1dad31R162)

https://github.com/grunk-dev/parametric/blob/cf920ff7f8cb30a2f527471fda19fc361e65549a/parametric/core.hpp#L149



It is not clear to me, how p gets created: the class param doesn't have a constructor that takes a shared_pointer as an argument. How does this work?

Yes, it does. It's private.

***In GitLab by @AntonReiswich on Sep 13, 2023, 14:59 UTC:***

Ahhh, yes!

**Comment by joergbrech** [comment]
mentioned in issue grunk#79

**Comment by joergbrech** [comment]
**Discussion in GitLab:**

***In GitLab by @AntonReiswich on Sep 18, 2023, 08:30 UTC:***

Commented on [parametric/core.hpp line 539](https://github.com/grunk-dev/parametric/compare/ea691e5c9fd9647fda2c2ea9a728d2fe1a130ed5..32d87343b88537e88f9e08a2d947f5b9082d6280#diff-092665cd611f3e8daaf379734eaf4a6a9e96a506bf09e52d584eeb76bc1dad31L539)

https://github.com/grunk-dev/parametric/blob/32d87343b88537e88f9e08a2d947f5b9082d6280/parametric/core.hpp#L539



What was the reason for deleting this class?

It was the custom compute node used by the `eval` function. There are new custom compute node classes now defined internally that use the new definition of `ComputeNode`.

**Comment by joergbrech** [comment]
**Discussion in GitLab:**

***In GitLab by @AntonReiswich on Sep 18, 2023, 08:47 UTC:***

Commented on [parametric/dag.hpp line 89](https://github.com/grunk-dev/parametric/compare/26df6e48dfb233966c25dd85da4b2de3bdf3592c..cf920ff7f8cb30a2f527471fda19fc361e65549a#diff-96aeefb0374ca00abc002562ed64c573e49b159ed8f4dc21d2f5557ff33746aeL89)

https://github.com/grunk-dev/parametric/blob/cf920ff7f8cb30a2f527471fda19fc361e65549a/parametric/dag.hpp#L90



I didn't understand the reason for removing this line - and how this helps to compute for example k*k?

Consider the expression `k*k` as a compute node, say `m`. Now, all inputs of a compute node will be connected to it automatically *(it is not the responsibility of the user anymore)*. To avoid duplicate invalidation of `m` if `k` changes, we make sure that `m` is not a duplicate child of `k`. Conversely, we use the `parents` vector to query the inputs, e.g. the first and second argument of `m`. Therefore we want to allow duplicate parents, but not children.

**Comment by joergbrech** [comment]
***In GitLab by @AntonReiswich on Sep 18, 2023, 13:04 UTC:***

approved this merge request

**Comment by joergbrech** [comment]
***In GitLab by @AntonReiswich on Sep 18, 2023, 13:04 UTC:***

mentioned in commit 712205edab3ed3e347b9efca13671d14dc35f957

**Comment by joergbrech** [comment]
mentioned in merge request grunk!35
